### ISSUE TITLE
`@bazel_tool//tools/cpp:link_extra_libs` should not propagate to exec config

### ISSUE THREAD

**Issue by tpudlik** [bazelbuild/bazel#22457]
### Description of the feature request:

The C++ compilation rules have a neat feature: they [allow you](https://bazel.build/reference/be/c-cpp#cc_binary.link_extra_lib) to specify a library that should be a linktime dependency of all `cc_binary` targets. That library is configured through a label flag, `@bazel_tool//tools/cpp:link_extra_libs`.

Unfortunately, this label flag persists across the transition from the target to the exec configuration. This is very inconvenient: if the `link_extra_libs` you want for the target platform include generated code, you run into circular dependencies when trying to build any binaries used in their generation. This tends to happen with [some Pigweed libraries](https://pigweed.dev/build_system.html#docs-build-system-bazel-link-extra-lib), preventing us from using `link_extra_libs` in most cases.

@gregestren I recall that internally we've done some work to restrict flag propagation to the exec config (b/292617118). What's the status on the open source side?

### Which category does this issue belong to?

Configurability

### What is the output of `bazel info release`?

release 7.1.2

**Comment by gregestren** [comment]
I'm sure this is just a matter of flipping https://github.com/bazelbuild/bazel/blob/4d2250561fc7b16b35eb99cb51921ec530457a4a/src/main/java/com/google/devtools/build/lib/analysis/config/CoreOptions.java#L125-L126.

We flipped that in the org-wide .bazelrc in b/292617118 but not by default.

I think it's reasonable to flip it here, combined with an announcement to bazel@head users on how to fix potential breakages. 

How about I put up a PR and we'll see if standard Bazel CI passes as a starting point?

**Comment by fmeum** [comment]
@gregestren I'm in favor of flipping it, but we should follow the incompatible change procedure.

Cc @meteorcloudy

**Comment by tjgq** [comment]
Is it a good idea to default `--experimental_exclude_starlark_flags_from_exec_config` to true? I think this is going to be extremely non-intuitive. What if the ability to propagate to `exec` was a property of the configuration flag, instead of determined by a separate command line option, i.e. `label_flag(..., propagate_to_exec = True)`? That way, it's harder to miss that you need to make an explicit decision when defining the flag.

**Comment by tpudlik** [comment]
Ah, I didn't realize that `--experimental_exclude_starlark_flags_from_exec_config` is just a flag that I can add to my `.bazelrc` to enable (pre-adopt) this behavior. I don't think it's documented anywhere! Let me enable this as a workaround.

As a user, I personally would like to see this flag flipped, but I haven't looked deeply into the pros and cons. What I _do_ feel strongly about (as a developer of a library that's built in both places) is that we should set the flag to the same default value both internally and in open-source Bazel!

**Comment by fmeum** [comment]
> Is it a good idea to default `--experimental_exclude_starlark_flags_from_exec_config` to true? I think this is going to be extremely non-intuitive. What if the ability to propagate to `exec` was a property of the configuration flag, instead of determined by a separate command line option, i.e. `label_flag(..., propagate_to_exec = True)`? That way, it's harder to miss that you need to make an explicit decision when defining the flag.

@gregestren What do you think about:
1. adding a `propagate_to_exec` flag to build settings that defaults to `True`.
2. turning `--experimental_exclude_starlark_flags_from_exec_config` into an incompatible flag that flips the default to `False`.
3. flipping the incompatible flag.

**Comment by gregestren** [comment]
@aranguyen has a proposal for a proper flag propagation model that they intend to implement in June. Which I think would cover 1. I'll see if I can share the proposal next week.

That sequence generally sounds fine.

**Comment by gregestren** [comment]
Update:

I'm now confident we should follow the process described in the last two comments. 

As this thread notes, the complication is if someone *needs* their Starlark flag to still propagate, you can only do that in a .`bazelrc`. Which doesn't generalize well.

> What I do feel strongly about (as a developer of a library that's built in both places) is that we should set the flag to the same default value both internally and in open-source Bazel!

Agreed. The current discrepancy is because Google has its global `.bazelc` that solves this problem. Bazel of course doesn't.

Here's the sequence:

1. Let @aranguyen 's [customizable flag propagation](https://github.com/bazelbuild/bazel/discussions/22997) support finish. This makes it possible for Bazel to understand and apply flag annotations like "only propagate to X".
1. Add a "propagate_to_exec" flag to build settings, as @fmeum notes
1. Rename / flip `--experimental_exclude_starlark_flags_from_exec_config`


1 is well underway. We'll follow up when the implementation lands.

**Comment by gregestren** [comment]
> How far out is propagate_to_exec? 

It'd be great to aim for October @ bazel head. 

> the flag cannot be controlled via the command-line.

Can you clarify this point?

**Comment by armandomontanez** [comment]
For reasons I haven't dug too deeply into, with `--experimental_exclude_starlark_flags_from_exec_config` enabled I am unable to flip the `--@rules_cc//cc/toolchains:experimental_enable_rule_based_toolchains` flag either from a `.bazelrc` or via specifying it as a `bazel buidl` CLI flag. I believe the root of this issue stems from the fact that the flag evaluation happens as a dependency of a `toolchain` rule, which does some number of transitions.

I tried allowlisting this flag by doing this in my .bazelrc:
```
build --experimental_propagate_custom_flag=@rules_cc//cc/toolchains:experimental_enable_rule_based_toolchains
build --@rules_cc//cc/toolchains:experimental_enable_rule_based_toolchains=True
```

But it still didn't take effect.

**Comment by gregestren** [comment]
@armandomontanez do you have a simple reproduction?

That doesn't sound like something `propagate_to_exec` would change, since that's just moving the source of truth vs. changing the underlying semantics.

**Comment by armandomontanez** [comment]
Also hitting this with `--@rules_rust//rust/toolchain/channel=nightly`, I suspect the reproducer is as easy as "make a toolchain that uses a flag of some kind, and then try to toggle it from `.bazelrc` or the CLI with `--experimental_exclude_starlark_flags_from_exec_config`"

I'll see if I can set aside time to make that reproducer.

**Comment by tpudlik** [comment]
It looks like the test for this behavior is excluded from CI, which is not a good sign: https://github.com/bazelbuild/bazel/blob/4e89bc299e6e655094c749700146ea7acca4a97c/src/test/java/com/google/devtools/build/lib/analysis/config/BuildConfigurationValueTest.java#L341-L390

**Comment by armandomontanez** [comment]
Reproducer at https://github.com/armandomontanez/bazel_reproducers/tree/main/exec_propagate_to_toolchain

TL;DR: I don't think this is a bug. It's just a limitation imposed by this flag.

**Comment by tpudlik** [comment]
Armando, shouldn't your reproducer set `--experimental_propagate_custom_flag`? I thought the bug was that this flag has no effect.

**Comment by armandomontanez** [comment]
Right you are, but that now seems to do the trick in my minimal example. Perhaps the issue I was hitting was unique to rules_cc toolchains, for a while they had special transition magic under the hood.

**Comment by armandomontanez** [comment]
I think I figured it out, it seems that `--experimental_propagate_custom_flag` must reference canonical repo names rather than apparent repo names.

**Comment by gregestren** [comment]
> It looks like the test for this behavior is excluded from CI, which is not a good sign:
> 
> [bazel/src/test/java/com/google/devtools/build/lib/analysis/config/BuildConfigurationValueTest.java](https://github.com/bazelbuild/bazel/blob/4e89bc299e6e655094c749700146ea7acca4a97c/src/test/java/com/google/devtools/build/lib/analysis/config/BuildConfigurationValueTest.java#L341-L390)
> 
> Lines 341 to 390 in [4e89bc2](/bazelbuild/bazel/commit/4e89bc299e6e655094c749700146ea7acca4a97c)
> 
>    @Test 
>    // TODO: b/377959266 - fix test setup bug that fails Bazel CI. This test's logic doesn't cause the 
>    //   bug: it's somehow caused by test naming. See bug for details. 
>    @Ignore("b/377959266") 
>    public void testExecStarlarkFlag_isPropagatedByTargetPattern() throws Exception { 

https://github.com/bazelbuild/bazel/pull/25259 re-instated that test. (We never figured out what the problem was, but it was 100% an issue with the test infrastructure vs. this test's logic).

**Comment by gregestren** [comment]
Update: The plan from https://github.com/bazelbuild/bazel/issues/22457#issuecomment-2237687535 remains. 

@aranguyen  has prototyped Starlark flag's controlling their exec propagation and is debugging a circular dependency issue. After resolving that they'll share a proposal we can all comment on. 

I expect a little debate on that proposal. Particularly that defaulting a flag's value on exec transitions is less general purpose than setting it to some arbitrary value. But supporting the latter requires more API work. We'll see what we're facing when we examine the proposal.

**Comment by tpudlik** [comment]
@gregestren is there a proposal or shareable prototype yet? I just created a new label flag in our project that should, in fact, propagate to the exec configuration (http://pwrev.dev/306081), so this came up again!

**Comment by gregestren** [comment]
I'll status-check this week.

**Comment by gregestren** [comment]
@aranguyen and I synced today. Technical prototype validated, proposal drafted and preparing for sharing.

**Comment by gregestren** [comment]
https://github.com/bazelbuild/bazel/issues/26909 planned for Bazel 9.0.
